src/Stratis.Features.FederatedPeg/CollateralFederationManager.cs

```diff
@@ -0,0 +1,68 @@
+﻿using System.Collections.Generic;
+using System.Linq;
+using Microsoft.Extensions.Logging;
+using NBitcoin;
+using Stratis.Bitcoin.Configuration;
+using Stratis.Bitcoin.Features.PoA;
+using Stratis.Bitcoin.Signals;
+using Stratis.Bitcoin.Utilities;
+
+namespace Stratis.Features.FederatedPeg
+{
+    public class CollateralFederationManager : FederationManager
```

Comment:
>and the caller wont know the difference from where it originated from.

That's the whole point of inheritance, abstractions and interfaces: not knowing what we are dealing with except for knowing the public interface. 

So imho it's not a bad thing, it's a tool. A tool which can be misused by mistake, true. 

>Ivan, if you want to push the PR like this, that is fine, but then please consider a design where the base logic sits in an abstract class and we have two inheriting classes that both implement abstract save and load methods. That way there is no room any error going forward.

That will result in having everything in abstract class and actual behaviors will just implement save and load logic. 

Not sure it's worth it. 
The problem that you've mentioned is solved by having a review. Once it's reviewed and is correct there is no room for mistake since the only mistake we can have is not overriding those methods when we intend to.

Reply:
@noescape00, I am of the opinion that it is preferable to make it as difficult as possible to introduce a mistake into the code. The difficulty of relying on a review to catch the sort of error that can arise here, is that the full inheritance tree is not generally readily apparent from the diff unless it is all being introduced simultaneously like it is in this instance.

If it was done in separate PRs, for example, they might be reviewed by different people, or reviewed far apart. This removes the context and makes it likely that someone will overlook a method needing to be overridden for correctness. I believe something like this occurred recently with the PoA behaviours, so it is not an unheard-of scenario.

Reply:
Yes, very good point, I agree.
I don't agree with the potential solution which would be never having a default implementation and keep the method abstract so we just can't forget to override smth since it will result in a compilation error.

Reply:
@noescape00 My outlook on designing classes like this is always to ensure that we limit the amount of risk we expose ourselves too as much as we can. Saying that we should rely on the reviewer  is very risky for obvious reasons as stated by Kevin.

Yes, a lot of the logic will be contained in the base class but so what? Moving the shared code to a base abstract class will take 5 minutes and then you never have to worry if the wrong save or load method will be called. Those ARE very important methods. I don't understand what the problem is with this approach. Why introduce code where mistakes can be made as opposed to code where the room for error is zero? We **just** had a problem in the POA code because of the inheriting nature of those behavior classes. If we implemented a base abstract class the chances for a mistake, like the one that was made, would have been prevented by design.

Reply:
Ok, fair point. 
Fixed

Reply:
I'm late for the discussion, I haven't worked on federatin code but design wise I share the same view of @zeptin2 / @fassadlr so +1 for abstract.
But well... you did it already :)